### PR TITLE
CORTX-29930: Enhance monitoring capability for control node

### PR DESCRIPTION
**Pull request by mariyappanp**
# Problem Statement
Add control node cluster cardinality and watch control node events

# Design
[CORTX-29930](https://jts.seagate.com/browse/CORTX-29930) - Enhance monitoring capability for control node

# Coding
-  [x] Coding conventions are followed and code is consistent

# Testing 
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

[test_results.txt](https://jts.seagate.com/secure/attachment/513113/test_results.txt)

# Review Checklist 
- [x] PR is self reviewed
- [x] JIRA number/GitHub Issue added to PR
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained
- [ ] Is there a change in filename/package/module or signature? [Y/N]: 
- [ ] If yes for above point, is a notification sent to all other cortx components? [Y/N]
- [ ] Side effects on other features (deployment/upgrade)? [Y/N]
- [ ] Dependencies on other component(s)? [Y/N]
-     If yes for above point, post link to the corresponding PR.

# Review Checklist 
- [ ] Is perfline test run and the report with and without the changes updated in the PR? [Y/N]: 

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: ha/k8s_setup/const.py [hunk 1]
```diff
@@ -43,10 +43,12 @@
 class GconfKeys(enum.Enum):
     NODE_CONST = "node"
     SERVICE_CONST = "services"
+    TYPE_CONST = "type"
```
Comment (mariyappanp) [inline ha/k8s_setup/const.py:46]: Duplicate. This PR has dependency with PR#683.

File: ha/k8s_setup/const.py [hunk 2]
```diff
@@ -43,10 +43,12 @@
...
     NAME_CONST = "name"
     CVG_NAME = "node{_DELIM}{node_id}{_DELIM}storage{_DELIM}cvg[{cvg_index}]{_DELIM}name"
     CVG_COUNT = "node{_DELIM}{node_id}{_DELIM}storage{_DELIM}num_cvg"
     DATA_COUNT = "node{_DELIM}{node_id}{_DELIM}storage{_DELIM}cvg[{cvg_index}]{_DELIM}devices{_DELIM}num_data"
     METADATA_COUNT = "node{_DELIM}{node_id}{_DELIM}storage{_DELIM}cvg[{cvg_index}]{_DELIM}devices{_DELIM}num_metadata"
     DATA_DISK = "node{_DELIM}{node_id}{_DELIM}storage{_DELIM}cvg[{cvg_index}]{_DELIM}devices{_DELIM}data[{d_index}]"
     METADATA_DISK = "node{_DELIM}{node_id}{_DELIM}storage{_DELIM}cvg[{cvg_index}]{_DELIM}devices{_DELIM}metadata[{m_index}]"
+    CONTROL_NODE = "control_node"
```
Comment (mariyappanp) [inline ha/k8s_setup/const.py:54]: Duplicate. This PR has dependency with PR#683.

File: ha/util/conf_store.py [hunk 3]
```diff
@@ -221,3 +222,30 @@ def get_disk_list_for_cvg(index, node_id, cvg_id):
...
+        node_id = []
+        try:
+            node_key = Conf.search(index, GconfKeys.NODE_CONST.value,
+                                   GconfKeys.TYPE_CONST.value, GconfKeys.CONTROL_NODE.value)
+            for key in node_key:
+                value = key.split('>')[1]
+                node_id.append(value)
+            Log.info(f"Fetched control node machine-id: {node_id}")
+        except Exception as e:
+            Log.error(f"Unable to fetch control node machine-id. Error:{e}")
```
Comment (mariyappanp) [inline ha/util/conf_store.py:250]: Duplicate. This PR has dependency with PR#683.

**Review by akash2144 — COMMENTED** [review]
cortx>ha>v1>cluster_cardinality:{"num_nodes": 8, "node_list": ["d5bd40a41da44c8dbf90ce418453f711", "645fe1d98a774b0a9b1a5972c6857741", "2434adbe55f046c2a3ec219d04b720d3", "cad486c5d73d44de9413ebaf8e16c6b8", "099d7bf619e24591a3f01b00324ce18a", "4c38e91086d745cf9104798a2e96785b", "926e2fb713924aa8a5b156695206364c", "dd8d2ceb64444e6caeb5532753306263", "04539dd8be544c428c56d9b22baec91a"]}

Please check why num_nodes count is mismatched. it should be 9.

File: ha/util/conf_store.py [hunk 4]
```diff
@@ -98,9 +98,10 @@ def set_cluster_cardinality(self, index):
 
         data_pods = ConftStoreSearch.get_data_pods(index)
         server_pods = ConftStoreSearch.get_server_pods(index)
+        control_pods = ConftStoreSearch.get_control_pods(index)
 
         # Combine the lists data_pods, server_pod and find unique machine ids
```
Comment (Madhura-08) [inline ha/util/conf_store.py:103]: Please update comment for control pod as well.

**Review by gauri-bhosale — APPROVED** [review]
please remove duplicate code after dependent PR has been merged.

**Comment by mariyappanp** [comment]
> cortx>ha>v1>cluster_cardinality:{"num_nodes": 8, "node_list": ["d5bd40a41da44c8dbf90ce418453f711", "645fe1d98a774b0a9b1a5972c6857741", "2434adbe55f046c2a3ec219d04b720d3", "cad486c5d73d44de9413ebaf8e16c6b8", "099d7bf619e24591a3f01b00324ce18a", "4c38e91086d745cf9104798a2e96785b", "926e2fb713924aa8a5b156695206364c", "dd8d2ceb64444e6caeb5532753306263", "04539dd8be544c428c56d9b22baec91a"]}
> 
> Please check why num_nodes count is mismatched. it should be 9.
> 
> Looks like data/key was already there. And again mini provision was performed. And consul update happened(consul kv put) which I think overwrites the data.
Initially, I have updated control node in cardinality and tested monitoring functionality. This test result file is not updated later.
Before posting this PR, I have tested thoroughly and count is 9. I will update the test result file. @mukhtar-inamdar

File: ha/util/conf_store.py (hunk 4 above)
Comment (Madhura-08) [reply ha/util/conf_store.py:103]: Yes. I got it. I am saying, for line 103, update the comment for control pod as well.

**Review by akash2144 — APPROVED** [review]

File: ha/util/conf_store.py (hunk 4 above)
Comment (mariyappanp) [reply ha/util/conf_store.py]: Done.

Comment (Madhura-08) [reply ha/util/conf_store.py]: Thanks.

**Review by Madhura-08 — APPROVED** [review]

**Pull request merged**